ls1intum/Artemis
# [Feature] More Flexible Format for Transmission of Test Results from Jenkins

### Checklist

### Motivation and Context
Closes #2861. Extensive description provided there.

### Description
This implements the changes outlined in the Jenkins-section of #2861.
PR for the Jenkins-Plugin: ls1intum/jenkins-server-notification-plugin/pull/8

#### Client-facing changes
Before this PR the client assumed that no messages for successful tests needed to be shown. This PR makes successful tasks in the exercise description clickable to show a detailed info, only if they have a message associated with them. The same applies to the result string (e.g. `10 of 10 passed`) at the top that becomes clickable only when messages exist. (c.f. Screenshots)


### Steps for Testing

The main changes of this PR apply only to Artemis+Gitlab+Jenkins setups and also **require a test server that runs an updated version of the Jenkins plugin** (ls1intum/jenkins-server-notification-plugin/pull/8).

Testing on TS2 with the Jenkins setup requires an update of the Jenkins plugin. Therefore, it is probably easiest for people with a local development setup of Artemis+Gitlab+Jenkins to test the actual functionality.
Alternative to anyone who has direct access to the actual Jenkins installation of TS2: The updated plugin is available as artefact [here](https://github.com/ls1intum/jenkins-server-notification-plugin/actions/runs/643218337). This should be installable just for the duration of the testing session and could then be reverted afterwards to the actual [release](https://github.com/ls1intum/jenkins-server-notification-plugin/releases).

However, it should be tested that the client changes do not show any messages that are not useful to students for successful tests when running in a Bamboo setup. The message format in the [Plugin’s readme](https://github.com/ls1intum/bamboo-server-notification-plugin) suggests that no messages for successful tests are present.

To see an effect for Jenkins setups a custom feedback has to be created while executing the tests. The easiest way to create a custom feedback file that can be parsed by the Jenkins Plugin:
1. Create a Python programming exercise.
2. In the `test` repository replace the top part of `behaviour/behaviour_test.py` to
```python
import unittest

from assignment.sorting_algorithms import *
from assignment.context import Context
from assignment.policy import Policy

import os
import json

def create_feedback(filename, name, successful, message, has_name=True, has_success=True, has_message=True):
    if not os.path.isdir("customFeedbacks"):
        os.mkdir("customFeedbacks")
    feedback = {}
    if has_name:
        feedback["name"] = name
    if has_success:
        feedback["successful"] = successful
    if has_message:
        feedback["message"] = message
    f = open(f"customFeedbacks/{filename}", "w")
    f.write(json.dumps(feedback))
    f.close()


class TestSortingBehavior(unittest.TestCase):
    unordered = []
    ordered = [1, 1, 2, 3, 4, 8]

    def setUp(self):
        self.unordered = [3, 4, 2, 1, 8, 1]

    def test_bubble_sort(self):
        # should be sent to Artemis, clickable in overview with message
        create_feedback("sucessAll.json", "successAll", True, "With message")

        # should be sent to Artemis, non-clickable as they are without message
        create_feedback("successNullMessage.json", "successNullMessage", True, None)
        create_feedback("successEmptyMessage.json", "successEmptyMessage", True, "     ")
        create_feedback("successNoMessage.json", "successNoMessage", True, None, has_message=False)

        # those two should not be sent to Artemis, as they are invalid, should only produce info message in Jenkins log
        create_feedback("missingName.json", "missingName", True, "msg", has_name=False)
        create_feedback("missingName2.json", "missingNameFailed", False, "msg", has_name=False)

        # success attribute missing => set to default value false and still sent to Artemis
        create_feedback("missingSuccess.json", "missingSuccess", True, "With message", has_success=False)

        bubble_sort = BubbleSort()
        bubble_sort.perform_sort(self.unordered)
        self.assertEqual(self.ordered, self.unordered)

    # test_merge_sort and everything below unchanged
```
This alters the test case to also create an additional custom feedback. This also shows that the new method of creating feedback is not really meant to be used with the provided templates but for custom build processes that use other tools.

3. Submit the change and alter the exercise description to also include the new test cases to be able to easily see the results:
```md
1. [task][missingSuccess](missingSuccess)
2. [task][successAll](successAll)
3. [task][successEmptyMessage](successEmptyMessage)
4. [task][successNoMessage](successNoMessage)
5. [task][successNullMessage](successNullMessage)
6. [task][missingName](missingName)
7. [task][missingNameFailed](missingNameFailed)
```
4. Add additional `create_feedback` calls to verify they behave as expected.
    The format of the custom feedbacks has been defined as:
    * `name`: This is the name of the test case as it will be shown for example on the ‘Configure Grading’ page.
      It should therefore have a for this exercise uniquely identifiable name and **has to be non-null**.
    * `successful`: Indicates if the test case execution for this submission should be marked as successful or failed. Defaults to `false` (= non-successful/failed) if missing.
    * `message`: The message shown as additional information to the student.
      **Required for non-sucessful feedback**, optional otherwise.


### Test Coverage
`JenkinsService.java`: 58% ⟹ 61% (in the changed function: 80% ⟹ 97%)
`FeedbackService.java`: 94% ⟹ 95%
`programming-exercise-instruction-task-status.component.ts`: 69% ⟹ 71%


### Screenshots
<details>
  <summary>Click to expand!</summary>

Overview when clicking on the result string: Successful feedbacks without messages are still grouped together, the one with a custom message is not:
![Screenshot_20210220_112522](https://user-images.githubusercontent.com/64250573/108592483-5c799200-736e-11eb-9069-40ccdf8d3879.png)


Exercise instruction after submission, `successAll` clickable and has message, the successful test cases without message are not clickable, the two saved in an invalid format have not been received:
![Screenshot_20210220_111822](https://user-images.githubusercontent.com/64250573/108592275-664ec580-736d-11eb-938c-b85010e7c024.png)


</details>
